### PR TITLE
Fix: wrong icons showing when selecting invisible units (Spectator/Ch…

### PR DESCRIPTION
**Pull request by Helwor**
…eat)

While it is not possible to retrieve the defID through spGetUnitDefID, spGetSelectedUnitsSorted can give them. By the way some optimization, saving multiple call of spGetUnitDefID aswell as the whole process widget:SelectionChanged()

### CODE REVIEW

**Comment by sprunk** [comment]
> While it is not possible to retrieve the defID through spGetUnitDefID, spGetSelectedUnitsSorted can give them

Sounds like engine bug if these functions are so inconsistent. And generally, the whole "selecting invisible units" behaviour is probably engine bug.

But if this patch also works as an optimisation for the normal case then I think we can take it. I'll test on the weekend.

**Comment by Helwor** [comment]
I don't know if I would consider a bug being able to select and control units in god mode just because not having global vision.  But I can tell that's pretty useful as dev to have this ability.

File: LuaUI/Widgets/gui_chili_selections_and_cursortip.lua [hunk 1]
```diff
@@ -1026,7 +1027,6 @@ local function SelectionsIconClick(button, unitID, unitList, unitDefID)
 			newSelectedUnits = selectedUnitsList
 		end
 		spSelectUnitArray(newSelectedUnits)
-		widget:SelectionChanged(newSelectedUnits)
```
Comment (sprunk) [inline LuaUI/Widgets/gui_chili_selections_and_cursortip.lua:1029]: This was always redundant and doesn't depend on the rest of the patch, correct?

File: LuaUI/Widgets/gui_chili_selections_and_cursortip.lua [hunk 2]
```diff
@@ -1904,6 +1906,7 @@ end
 
 local function GetSingleUnitInfoPanel(parentControl, isTooltipVersion)
 	local selectedUnitID
+	local selectedUnitDefID
```
Comment (sprunk) [inline LuaUI/Widgets/gui_chili_selections_and_cursortip.lua:1909]: Unused?

File: LuaUI/Widgets/gui_chili_selections_and_cursortip.lua [hunk 3]
```diff
@@ -2806,7 +2809,18 @@ function widget:KeyRelease(key, modifier, isRepeat)
 	end
 end
 
-function widget:SelectionChanged(newSelection)
+function widget:CommandsChanged()
```
Comment (sprunk) [inline LuaUI/Widgets/gui_chili_selections_and_cursortip.lua:2812]: Why change from `SelectionChanged` to `CommandsChanged`?

**Comment by Helwor** [comment]
hi, so SelectionChanged is used for modifying the selection on the fly by returning an array, and if it happens, SelectionChanged will be immediately called again. That's why I was saying in the description "aswell as the whole process widget:SelectionChanged()". In the opposite, CommandsChanged is called only once when the selection is finally made. You can check that by right clicking some units of your selection IIRC, SelectionChanged will be called twice. So using CommandsChanged prevent redundancies of the whole process.

**Comment by sprunk** [comment]
Cool, thanks!

**Pull request merged**